File: app/domain/projects/aspect_ratio_impact.py
"""切换项目画幅前的影响披露：只读接口，不改任何数据。

冻结契约（主会话 2026-09-23 与前端并行开发约定，不得另起形状）：

``GET /api/projects/{project_id}/aspect-ratio-impact?aspect_ratio=16:9`` 返回::

    {"project_id": str, "current_aspect_ratio": "9:16"|"16:9",
     "target_aspect_ratio": "9:16"|"16:9", "adopted_videos_total": int,
     "adopted_videos_mismatched": int, "scene_images_total": int,
     "scene_images_mismatched": int | null}

版本 meta 没有画幅快照的一律按 "9:16" 计（上线前生产全是 9:16）；场景图
（``scene_references``）没有存尺寸/画幅记录，``scene_images_mismatched``
如实返回 ``null``，不猜测——前端据此提示「场景图 N 张，画幅未记录」。
"""
from __future__ import annotations

from fastapi import HTTPException

from app.db import get_conn
from app.domain.common import _project_or_404, router
from app.project_settings import ASPECT_RATIOS, resolve_aspect_ratio


@router.get("/projects/{project_id}/aspect-ratio-impact")
def get_project_aspect_ratio_impact(project_id: str, aspect_ratio: str):
    """只读评估：切到 ``aspect_ratio`` 会让当前项目里多少已采用视频/场景图画幅不同。"""
    _project_or_404(project_id)  # 归属校验，与本包既有只读端点同一写法（见 listing.py）
    target = str(aspect_ratio or "").strip()
    if target not in ASPECT_RATIOS:
        raise HTTPException(422, f"不支持的画幅：{aspect_ratio!r}")
    conn = get_conn()
    current = resolve_aspect_ratio(conn, project_id)
    video_row = conn.execute(
        """SELECT COUNT(*) AS total,
                  SUM(CASE WHEN COALESCE(json_extract(v.image_inputs,'$.aspect_ratio'),'9:16')<>?
                           THEN 1 ELSE 0 END) AS mismatched
             FROM shots s
             JOIN episodes e ON e.id = s.episode_id
             JOIN shot_versions v ON v.id = s.adopted_version_id
            WHERE e.project_id = ? AND s.adopted_version_id IS NOT NULL""",
        (target, project_id),
    ).fetchone()
    scene_total = conn.execute(
        "SELECT COUNT(*) AS c FROM scene_references WHERE project_id=?", (project_id,),
    ).fetchone()["c"]
    return {
        "project_id": project_id,
        "current_aspect_ratio": current,
        "target_aspect_ratio": target,
        "adopted_videos_total": int(video_row["total"] or 0),
        "adopted_videos_mismatched": int(video_row["mismatched"] or 0),
        "scene_images_total": int(scene_total or 0),
        # scene_references 没有存尺寸/画幅字段，如实返回"未记录"而不是猜测。
        "scene_images_mismatched": None,
    }

File: frontend/src/lib/aspectRatioImpact.ts
import type { AspectRatioImpact } from '../api'

const ASPECT_RATIO_LABEL: Record<string, string> = {
  '9:16': '9:16（竖屏）',
  '16:9': '16:9（横屏）',
}

export function aspectRatioLabel(value: string): string {
  return ASPECT_RATIO_LABEL[value] ?? value
}

export type AspectRatioImpactDialogContent = {
  title: string
  summary: string
  message: string
  details: string[]
  danger: boolean
}

/**
 * 画幅切换前确认弹窗的文案（2026-09-23 用户拍板）：如实列出已采用视频与场景图
 * 受影响的数量；impact 接口 404/失败时（impact 为 null 且 impactError 非空）
 * 照样允许切换，只是把"统计不到"如实说清楚——不能因为统计接口挂了就拦死整个
 * 切换（CLAUDE.md「拦住用户时必须给出路」）。纯函数，不碰组件状态，便于独立测试
 * 「mismatched=null」与「impact 失败」这两种形态而不必渲染整个面板。
 */
export function buildAspectRatioImpactDialog(
  target: string,
  impact: AspectRatioImpact | null,
  impactError: string | null,
): AspectRatioImpactDialogContent {
  const title = `切换画幅为 ${aspectRatioLabel(target)}？`
  if (!impact) {
    return {
      title,
      summary: impactError ? '暂时无法统计受影响的视频数' : '正在核对受影响的视频与场景图…',
      message: impactError
        ? `统计接口未能返回结果（${impactError}）；仍可以切换画幅，已采用的视频若是旧画幅，会在成片合成时被等比放大裁切进新画布，可能裁掉部分画面。`
        : '仍可以切换画幅；已采用的视频若是旧画幅，会在成片合成时被等比放大裁切进新画布，可能裁掉部分画面。',
      details: ['之后新生成的视频和场景图按新画幅', '需要的话可以到生成台重新生成'],
      danger: false,
    }
  }
  const sceneLine = impact.scene_images_mismatched == null
    ? `场景图 ${impact.scene_images_total} 张，画幅未记录`
    : `场景图 ${impact.scene_images_mismatched} 张为旧画幅（共 ${impact.scene_images_total} 张）`
  return {
    title,
    summary: `已采用的 ${impact.adopted_videos_mismatched} 个视频仍是旧画幅，成片合成时会等比放大裁切，可能裁掉画面`,
    message: `当前项目共采用 ${impact.adopted_videos_total} 个视频（当前画幅 ${aspectRatioLabel(impact.current_aspect_ratio)}）。`,
    details: [sceneLine, '之后新生成的按新画幅', '需要的话到生成台重新生成'],
    danger: impact.adopted_videos_mismatched > 0,
  }
}
